model/common/src/icon4py/model/common/interpolation/rbf_interpolation.py

```diff
@@ -0,0 +1,520 @@
+    domain = element_center_lat.domain
+    dim = domain[0].dim
+    zonal_meridional_f = _get_zonal_meridional_f(dim)
+
+    z_nx = []
+    nxnx = []
+    rhs = []
+
+    assert len(u) == len(v)
+    assert 1 <= len(u) <= 2
+    for i in range(len(u)):
+        z_nx_x = gtx.zeros(domain, dtype=ta.wpfloat, allocator=backend)
+        z_nx_y = gtx.zeros(domain, dtype=ta.wpfloat, allocator=backend)
+        z_nx_z = gtx.zeros(domain, dtype=ta.wpfloat, allocator=backend)
+
+        zonal_meridional_f(
+            element_center_lat,
+            element_center_lon,
+            u[i],
+            v[i],
+            out=(z_nx_x, z_nx_y, z_nx_z),
+            offset_provider={},
+        )
+        z_nx.append(array_ns.stack((z_nx_x.ndarray, z_nx_y.ndarray, z_nx_z.ndarray), axis=-1))
+        assert z_nx[i].shape == (rbf_offset.shape[0], 3)
+
+        nxnx.append(
+            array_ns.matmul(z_nx[i][:, array_ns.newaxis], edge_normal.transpose(0, 2, 1)).squeeze()
+        )
+        rhs.append(rbf_val * nxnx[i])
+        assert rhs[i].shape == rbf_offset.shape
+
+    # Compute dot product of normal vectors for RBF interpolation matrix
+    z_nxprod = _dot_product(edge_normal, edge_normal, array_ns=array_ns)
+    assert z_nxprod.shape == (
+        rbf_offset.shape[0],
+        rbf_offset.shape[1],
+        rbf_offset.shape[1],
+    )
+
+    # Distance between edge midpoints for RBF interpolation matrix
+    z_dist = _arc_length_pairwise(edge_center, array_ns=array_ns)
+    assert z_dist.shape == (
+        rbf_offset.shape[0],
+        rbf_offset.shape[1],
+        rbf_offset.shape[1],
+    )
+
+    # Set up RBF interpolation matrix
+    z_rbfmat = z_nxprod * _kernel(rbf_kernel, z_dist, scale_factor, array_ns=array_ns)
+    assert z_rbfmat.shape == (
+        rbf_offset.shape[0],
+        rbf_offset.shape[1],
+        rbf_offset.shape[1],
+    )
+
+    # Solve linear system for coefficients
+    #
+    # Currently always on CPU. At the time of writing cupy does not have
+    # cho_solve with the same interface as scipy, but one has been proposed:
+    # https://github.com/cupy/cupy/pull/9116.
+    # TODO: vectorize this?
+    rbf_vec_coeff_np = [np.zeros(rbf_offset.shape, dtype=ta.wpfloat) for j in range(len(u))]
+    rbf_offset_np = data_alloc.as_numpy(rbf_offset)
+    z_rbfmat_np = data_alloc.as_numpy(z_rbfmat)
+    rhs_np = [data_alloc.as_numpy(x) for x in rhs]
+    for i in range(z_rbfmat.shape[0]):
+        invalid_neighbors = np.where(rbf_offset[i, :] < 0)[0]
+        num_neighbors = rbf_offset_np.shape[1] - invalid_neighbors.size
+        rbfmat_np = z_rbfmat_np[i, :num_neighbors, :num_neighbors]
+        z_diag_np = sla.cho_factor(rbfmat_np)
+        for j in range(len(u)):
+            rbf_vec_coeff_np[j][i, :num_neighbors] = sla.cho_solve(
+                z_diag_np, np.nan_to_num(rhs_np[j][i, :num_neighbors])
+            )
+    rbf_vec_coeff = [array_ns.asarray(x) for x in rbf_vec_coeff_np]
```

Comment:
Note, with a GPU backend this part is still done on CPU (with scipy) because cupy doesn't have the `cho_solve` API exposed, though one is proposed in a PR. Unless this becomes a performance bottleneck, I would keep this on CPU for now, and potentially move it over to the GPU if/when cupy gets the same `cho_solve` API. One can compose the functionality of `cho_solve` from other parts in the cupy API, but I don't see it as worth the effort right now.